File: kotlin/kotlin-gettingstarted/src/main/kotlin/minipractices/taxipark/TaxiPark.kt
package minipractices.taxipark

data class TaxiPark(
        val allDrivers: Set<Driver>,
        val allPassengers: Set<Passenger>,
        val trips: List<Trip>)

abstract class RegisteredUser {
    abstract val firstName: String
    abstract val lastName: String
    abstract val age: Int
    abstract val id: Int
}

//data class Driver(val firstName: String, val lastName: String, val age: Int, val id: Int)
data class Driver(
    override val firstName: String,
    override val lastName: String,
    override val age: Int,
    override val  id: Int
): RegisteredUser()


//data class Passenger(val firstName: String, val lastName: String, val age: Int, val id: Int)
data class Passenger(
    override val firstName: String,
    override val lastName: String,
    override val age: Int,
    override val  id: Int
): RegisteredUser()


data class Trip(
        val driver: Driver,
        val passengers: Set<Passenger>,
        // the trip duration in minutes
        val duration: Int,
        // the trip distance in km
        val distance: Double,
        // the percentage of discount (in 0.0..1.0 if not null)
        val discount: Double? = null
) {
    // the total cost of the trip
    val cost: Double
        get() = (1 - (discount ?: 0.0)) * (duration + distance)
}

File: scala/scala2/scala2_gettingstarted/src/main/scala/minipractices/taxipark/TaxiPark.scala
package minipractices.taxipark

import minipractices.taxipark.RegisteredUserType.RegisteredUserType

 abstract class RegisteredUser {
  val firstName: String
  val lastName: String
  val age: Int
  val id: Int
}

object RegisteredUser {
  def create(userType: RegisteredUserType,
             firstName: String,
             lastName: String,
             age: Int,
             id: Int): RegisteredUser = {
    userType match {
      case RegisteredUserType.DRIVER => Driver(firstName, lastName, age, id)
      case RegisteredUserType.PASSENGER => Passenger(firstName, lastName, age, id)
    }
  }
}


case class Driver(
   override val firstName: String,
   override val lastName: String,
   override val age: Int,
   override val  id: Int) extends RegisteredUser{
}


case class Passenger(
   override val firstName: String,
   override val lastName: String,
   override val age: Int,
   override val  id: Int) extends RegisteredUser {
}


// https://www.baeldung.com/scala/enumerations
object RegisteredUserType extends Enumeration {
  type RegisteredUserType = Value
  val PASSENGER, DRIVER = Value
}


case class Trip (
    driver: Driver,
    passengers: Set[Passenger],
    duration: Int,
    distance: Double,
    discount: Double) {
  val cost: Double = (1 - discount)*(duration + distance)
}


case class TaxiPark(
   allDrivers: Set[Driver],
   allPassengers: Set[Passenger],
   trips: List[Trip])


case class TripWrapper (
     driver: Driver,
     passengers: Array[Passenger],
     duration: Int,
     distance: Double,
     discount: Double) {
      val cost: Double = (1 - discount) * (duration + distance)
}


case class TaxiParkWrapper (
   allDrivers: Array[Driver],
   allPassengers: Array[Passenger],
   trips: Array[TripWrapper])

